File: kotlin/dnltsk/src/main/kotlin/BeerSong.kt
class BeerSong {

    companion object {

        fun lyrics() = verses(99, 0)

        fun verses(to: Int, from: Int): String {
            return (from..to)
                    .reversed()
                    .map({ verse(it) })
                    .joinToString("")
        }

        fun verse(bottleCount: Int): String {
            return when (bottleCount) {
                0 -> "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n\n"
                1 -> "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n\n"
                2 -> "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n\n"
                else -> "$bottleCount bottles of beer on the wall, $bottleCount bottles of beer.\nTake one down and pass it around, ${bottleCount - 1} bottles of beer on the wall.\n\n"
            }
        }


    }


}

File: ruby/mediafinger/beer_song.rb
class BeerSong
  def lyrics
    verses(99, 0)
  end

  def verses(higher, lower)
    song = (lower..higher).reverse_each.map do |bottles|
      verse(bottles)
    end
    song.join("\n")
  end

  def verse(bottles)
    case bottles
    when 1
      <<~TEXT
        1 bottle of beer on the wall, 1 bottle of beer.
        Take it down and pass it around, no more bottles of beer on the wall.
      TEXT
    when 0
      <<~TEXT
        No more bottles of beer on the wall, no more bottles of beer.
        Go to the store and buy some more, 99 bottles of beer on the wall.
      TEXT
    else
      <<~TEXT
        #{bottles} #{pluralize('bottle', bottles)} of beer on the wall, #{bottles} #{pluralize('bottle', bottles)} of beer.
        Take one down and pass it around, #{bottles - 1} #{pluralize('bottle', bottles - 1)} of beer on the wall.
      TEXT
    end
  end

  def pluralize(string, amount)
    return string if amount == 1
    "#{string}s"
  end
end
